src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs

```diff
@@ -0,0 +1,237 @@
+                var initializerExpression = GetInitializerExpression(initializer);
+
+                // default(type)
+                if (initializerExpression.IsKind(SyntaxKind.DefaultExpression))
+                {
+                    return true;
+                }
+
+                // literals, use var if options allow usage here.
+                if (initializerExpression.IsAnyLiteralExpression())
+                {
+                    return stylePreferences.HasFlag(TypingStyles.VarForIntrinsic);
+                }
+
+                // constructor invocations cases:
+                //      = new type();
+                if (initializerExpression.IsKind(SyntaxKind.ObjectCreationExpression) &&
+                    !initializerExpression.IsKind(SyntaxKind.AnonymousObjectCreationExpression))
+                {
+                    return true;
+                }
+
+                // explicit conversion cases: 
+                //      (type)expr, expr is type, expr as type
+                if (initializerExpression.IsKind(SyntaxKind.CastExpression) ||
+                    initializerExpression.IsKind(SyntaxKind.IsExpression) ||
+                    initializerExpression.IsKind(SyntaxKind.AsExpression))
+                {
+                    return true;
+                }
+
+                // other Conversion cases:
+                //      a. conversion with helpers like: int.Parse, TextSpan.From methods 
+                //      b. types that implement IConvertible and then invoking .ToType()
+                //      c. System.Convert.Totype()
+                var declaredTypeSymbol = semanticModel.GetTypeInfo(variableDeclaration.Type, cancellationToken).Type;
+                var expressionOnRightSide = initializerExpression.WalkDownParentheses();
+
+                var memberName = expressionOnRightSide.GetRightmostName();
+                if (memberName == null)
+                {
+                    return false;
+                }
+
+                var methodSymbol = semanticModel.GetSymbolInfo(memberName, cancellationToken).Symbol as IMethodSymbol;
+                if (methodSymbol == null)
+                {
+                    return false;
+                }
+
+                if (memberName.IsRightSideOfDot())
+                {
+                    var typeName = memberName.GetLeftSideOfDot();
+                    return IsPossibleCreationOrConversionMethod(methodSymbol, declaredTypeSymbol, semanticModel, typeName, cancellationToken);
+                }
+
+                return false;
+            }
+
+            private bool IsIntrinsicType(SyntaxNode declarationStatement) =>
+                declarationStatement.IsKind(SyntaxKind.VariableDeclaration)
+                ? ((VariableDeclarationSyntax)declarationStatement).Variables.Single().Initializer.Value.IsAnyLiteralExpression()
+                : false;
+
+            private ExpressionSyntax GetInitializerExpression(EqualsValueClauseSyntax initializer) =>
```

Comment:
This is the same type used by the compiler for both Checked and Unchecked expressions, the keyword alone is different. Parenthesis don't matter here, as I'm only trying to infer the type of the expression